README.md
# Lockit login

[![Build Status](https://travis-ci.org/zemirco/lockit-login.svg?branch=master)](https://travis-ci.org/zemirco/lockit-login)
[![NPM version](https://badge.fury.io/js/lockit-login.svg)](http://badge.fury.io/js/lockit-login)
[![Dependency Status](https://david-dm.org/zemirco/lockit-login.svg)](https://david-dm.org/zemirco/lockit-login)

Log in users to your Express app. The module is part of [Lockit](https://github.com/zemirco/lockit).

## Installation

`npm install lockit-login`

```js
var Login = require('lockit-login');
var utils = require('lockit-utils');
var config = require('./config.js');

var db = utils.getDatabase(config);
var adapter = require(db.adapter)(config);

var app = express();

// express settings
// ...
// sessions are required - either cookie or some sort of db
app.use(cookieParser());
app.use(cookieSession({
  secret: 'this is my super secret string'
}));

// create new Login instance
var login = new Login(config, adapter);

// use login.router with your app
app.use(login.router);

// listen to events [optional]
login.on('login', function(user, res, target) {
  res.send('Welcome ' + user.name);
})
```

## Configuration

More about configuration at [Lockit](https://github.com/zemirco/lockit).

## Features

- two-factor authentication
- track failed log in attempts
- lock account after too many failed login attempts
- track time and ip of log ins
- redirect unauthorized users to /login and save requested url to session
- input validation
- allow login with username and/or email

## Routes included

 - GET /login
 - POST /login
 - POST /login/two-factor
 - GET /logout

## REST API

If you've set `exports.rest` in your `config.js` the module behaves as follows.

 - all routes have `/rest` prepended
 - `GET /rest/login` is `next()`ed and you can catch `/login` on the client
 - `POST /rest/login` stays the same but sends JSON
 - `POST /rest/login/two-factor` stays the same but sends JSON
 - `GET /rest/logout` sends JSON and you can catch `/logout` on the client

## Test

`make test`

## License

MIT


History.md

##### 1.1.6 / 2015-07-23

- add eslint
- fix #8

##### 1.1.5 / 2015-07-06

- update dependencies

##### 1.1.4 / 2015-07-03

- update dependencies
- use eslint
- use make instead grunt

##### 1.1.3 / 2014-08-18

- escape query parameters for redirect
- update dependencies

##### 1.1.2 / 2014-05-27

- set `autocomplete="off"` in forms
- use Bootstrap responsive classes
- stop creating an extra test db for Travis
- use Bootstrap CSS in test app

##### 1.1.1 / 2014-05-23

- update lockit-utils

##### 1.1.0 / 2014-05-22

- add two-factor authentication
- add custom two-factor route via `config.login.twoFactorRoute`
- use `req.query.redirect` instead of adding redirect target to session
- make `action` in `get-login.jade` configurable
- use `util.destroy` helper method for logout

##### 1.0.1 / 2014-05-19

- set `req.session.loggedIn` to `true` on login
- code refactoring
- update dependencies

##### 1.0.0 / 2014-04-19

- requires Express 4.x
- makes use of `express.Router()`. No need to pass `app` around as argument.

  **old**

  ```js
  var Login = require('lockit-login');

  var login = new Login(app, config, adapter);
  ```

  **new**

  ```js
  var Login = require('lockit-login');

  var login = new Login(config, adapter);
  app.use(login.router);
  ```

  Listening on events **stays the same**.

  ```js
  login.on('login', function(user, res, target) {
    res.send('Welcome ' + user.name);
  })
  ```

- proper Error handling. All Errors are piped to next middleware.

  **old**

  ```js
  if (err) console.log(err);
  ```

  **new**

  ```js
  if (err) return next(err);
  ```

  Make sure you have some sort of error handling middleware at the end of your
  routes (is included by default in Express 4.x apps if you use the `express-generator`).

##### 0.7.0 / 2014-04-11

- `username` becomes `name`


test/app.js

/**
 * Module dependencies.
 */

var express = require('express');
var routes = require('./routes');
var user = require('./routes/user');
var http = require('http');
var path = require('path');

var login = require('../index.js');

function start(config) {

  config = config || require('./config.js');

  var app = express();

  // set basedir so views can properly extend layout.jade
  app.locals.basedir = __dirname + '/views';

// all environments
  app.set('port', process.env.PORT || config.port || 3000);
  app.set('views', __dirname + '/views');
  app.set('view engine', 'jade');
  // make JSON output simpler for testing
  app.set('json spaces', 0);
  app.use(express.favicon());
  app.use(express.bodyParser());
  app.use(express.methodOverride());
  app.use(express.cookieParser('your secret here'));
  app.use(express.cookieSession());

  // use forgot password middleware with testing options
  login(app, config);

  app.use(app.router);
  app.use(express.static(path.join(__dirname, 'public')));

// development only
  if ('development' == app.get('env')) {
    app.use(express.errorHandler());
  }

  app.get('/', routes.index);
  app.get('/users', user.list);

  // restrict routes to logged in users
  function restrict(req, res, next) {
    if (req.session.email && req.session.username) {
      next();
    } else {
      // redirect to login page but save url the user really wanted to visit
      var url = req.url;
      // save original url to session
      req.session.redirectUrlAfterLogin = url;
      res.redirect('/login');
    }
  }

  // dummy route for testing redirection after login
  app.get('/test', restrict, function(req, res) {
    res.send('well done');
  });

  http.createServer(app).listen(app.get('port'), function(){
    console.log('Express server listening on port ' + app.get('port'));
  });

  return app;

}

// export app for testing
if(require.main === module){
  // called directly
  start();
} else {
  // required as a module -> from test file
  module.exports = function(config) {
    return start(config);
  };
}

test/csrf.js

var request = require('supertest');
var should = require('should');
var utls = require('lockit-utils');
var cookie = require('cookie');

var config = require('./app/config.js');
var app = require('./app/app.js');

var db = utls.getDatabase(config);
var adapter = require(db.adapter)(config);

var _config = JSON.parse(JSON.stringify(config));
_config.csrf = true;
_config.port = 6000;
var _app = app(_config);

describe('# csrf', function() {

  before(function(done) {
    // create test user
    adapter.save('csrf', 'csrf@email.com', 'password', function(err, user) {
      // verify email for csrf
      adapter.find('name', 'csrf', function(err, user) {
        user.emailVerified = true;
        user.twoFactorEnabled = true;
        // save updated user to db
        adapter.update(user, function(err, user) {
          done();
        });
      });
    });
  });

  it('should include csrf token in "get-login" view', function(done) {
    request(_app)
      .get('/login')
      .end(function(err, res) {
        var cookies = cookie.parse(res.headers['set-cookie'][0]);
        var token = cookies.csrf;
        res.text.should.containEql('name="_csrf" value="' + token + '"');
        res.statusCode.should.equal(200);
        done();
      });
  });

  it('should include csrf token in "two-factor" view', function(done) {
    var agent = request.agent('http://localhost:6000');
    agent
      .get('/login')
      .end(function(err, res) {
        var cookies = cookie.parse(res.headers['set-cookie'][0]);
        var token = cookies.csrf;
        // give agent some time to put cookies from suitcase to store
        process.nextTick(function() {
          agent
            .post('/login')
            .set('x-csrf-token', token)
            .send({login: 'csrf', password: 'password'})
            .end(function(err, res) {
              res.statusCode.should.equal(200);
              done();
            });
        });
      });
  });

  after(function(done) {
    adapter.remove('csrf', done);
  });

});


test/rest.js

var request = require('supertest');
var should = require('should');
var superagent = require('superagent');
var utils = require('lockit-utils');
var totp = require('notp').totp;

var config = require('./app/config.js');
var app = require('./app/app.js');

var db = utils.getDatabase(config);
var adapter = require(db.adapter)(config);

// create third app for testing REST
var _config = JSON.parse(JSON.stringify(config));
_config.rest = true;
_config.port = 5000;
var _app = app(_config);

describe('# with REST enabled', function() {

  before(function(done) {
    // create a user whose email is not verified
    adapter.save('bopp', 'bopp@email.com', 'password', function() {
      // create another user with verified email that does not interfere with login attempts
      adapter.save('boat', 'boat@email.com', 'password', function(err, user) {
        // verify email for boat
        adapter.find('name', 'boat', function(err, user) {
          user.emailVerified = true;
          // save updated user to db
          adapter.update(user, function(err, user) {
            // create another user with email verified
            adapter.save('beep', 'beep@email.com', 'password', function(err, user) {
              // verify email for beep
              adapter.find('name', 'beep', function(err, user) {
                user.emailVerified = true;
                // save updated user to db
                adapter.update(user, function(err, user) {
                  // create another user for two factor auth
                  adapter.save('three', 'three@email.com', 'password', function(err, user) {
                    // verify email for three and activate two-factor
                    adapter.find('name', 'three', function(err, user) {
                      user.emailVerified = true;
                      user.twoFactorEnabled = true;
                      user.twoFactorKey = 'qwertz';
                      // save updated user to db
                      adapter.update(user, function(err, user) {
                        done();
                      });
                    });
                  });
                });
              });
            });
          });
        });
      });
    });
  });

  describe('GET /login', function() {

    it('should not catch the route', function(done) {
      request(_app)
        .get('/rest/login')
        .end(function(err, res) {
          res.text.should.containEql('Cannot GET /rest/login');
          done();
        });
    });

  });

  describe('POST /login', function() {

    it('should send a JSON error message when login field is invalid', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: '', password: 'secret'})
        .end(function(err, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Please enter your email/username and password"}');
          done();
        });
    });

    it('should send a JSON error message when password field is empty', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: 'user', password: ''})
        .end(function(err, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Please enter your email/username and password"}');
          done();
        });
    });

    it('should send a JSON error message when email is not verified', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: 'bopp', password: 'password'})
        .end(function(err, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Invalid user or password"}');
          done();
        });
    });

    it('should render a JSON error message when user is not in db', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: 'afkl', password: 'password'})
        .end(function(err, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Invalid user or password"}');
          done();
        });
    });

    it('should render a JSON error message when password is false', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: 'boat', password: 'something'})
        .end(function(err, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Invalid user or password"}');
          done();
        });
    });

    it('should show a JSON error after three failed login attempts', function(done) {
      // don't use "steve" again as account will get locked
      request(_app)
        .post('/rest/login')
        .send({login: 'beep', password: 'wrong'})
        .end(function(err, res) {
          request(_app)
            .post('/rest/login')
            .send({login: 'beep', password: 'wrong'})
            .end(function(err, res) {
              request(_app)
                .post('/rest/login')
                .send({login: 'beep', password: 'wrong'})
                .end(function(err, res) {
                  res.statusCode.should.equal(403);
                  res.text.should.equal('{"error":"Invalid user or password. Your account will be locked soon."}');
                  done();
                });
            });
        });
    });

    it('should lock the account after five failed login attempts', function(done) {
      // two more login attempts
      request(_app)
        .post('/rest/login')
        .send({login: 'beep', password: 'wrong'})
        .end(function(err, res) {
          request(_app)
            .post('/rest/login')
            .send({login: 'beep', password: 'wrong'})
            .end(function(err, res) {
              res.statusCode.should.equal(403);
              res.text.should.equal('{"error":"Invalid user or password. Your account is now locked for ' + config.accountLockedTime + '"}');
              done();
            });
        });
    });

    it('should not allow login with a locked user account', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: 'beep', password: 'password'})
        .end(function(err, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"The account is temporarily locked"}');
          done();
        });
    });

    it('should enable a locked account after certain amount of time', function(done) {
      setTimeout(function() {
        request(_app)
          .post('/rest/login')
          .send({login: 'beep', password: 'password'})
          .end(function(err, res) {
            res.statusCode.should.equal(204);
            done();
          });
      }, 2000);
    });

    it('should allow login in with a name', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: 'beep', password: 'password'})
        .end(function(err, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });

    it('should allow login in with an email', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: 'beep@email.com', password: 'password'})
        .end(function(err, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });

    it('should send a notice when two-factor auth is enabled', function(done) {
      request(_app)
        .post('/rest/login')
        .send({login: 'three', password: 'password'})
        .end(function(err, res) {
          res.text.should.equal('{"twoFactorEnabled":true}');
          done();
        });
    });

  });

  describe('POST /login/two-factor', function() {

    it('should send an error when token is invalid', function(done) {
      var agent = superagent.agent();
      // login
      agent
        .post('http://localhost:5000/rest/login')
        .send({login: 'three', password: 'password'})
        .end(function(err, res) {
          process.nextTick(function() {
            // enter invalid token
            agent
              .post('http://localhost:5000/rest/login/two-factor')
              .send({token: '123456'})
              .end(function(err, res) {
                res.statusCode.should.equal(401);
                done();
              });
          });
        });
    });

    it('should send a success message when token is valid', function(done) {
      var agent = superagent.agent();
      // login
      agent
        .post('http://localhost:5000/rest/login')
        .send({login: 'three', password: 'password'})
        .end(function(err, res) {
          process.nextTick(function() {
            // enter valid token
            var token = totp.gen('qwertz', {});
            agent
              .post('http://localhost:5000/rest/login/two-factor')
              .send({token: token})
              .end(function(err, res) {
                res.statusCode.should.equal(204);
                done();
              });
          });
        });
    });

  });

  describe('GET /logout', function() {

    var agent = superagent.agent();

    it('should only send the status code', function(done) {
      // login first
      agent
        .post('http://localhost:4000/login')
        .send({login:'beep', password:'password'})
        .end(function(err, res) {
          // then logout
          agent
            .get('http://localhost:5000/rest/logout')
            .end(function(err, res) {
              res.statusCode.should.equal(204);
              done();
            });
        });
    });

  });

  // remove user from db
  after(function(done) {
    adapter.remove('beep', function() {
      adapter.remove('boat', function() {
        adapter.remove('bopp', function() {
          adapter.remove('three', done);
        });
      });
    });
  });

});


index.js
'use strict';

var path = require('path');
var events = require('events');
var util = require('util');
var express = require('express');
var ms = require('ms');
var moment = require('moment');
var utils = require('lockit-utils');
var pwd = require('couch-pwd');

/**
 * Internal helper functions
 */

function join(view) {
  return path.join(__dirname, 'views', view);
}



/**
 * Login constructor function.
 *
 * @constructor
 * @param {Object} config
 * @param {Object} adapter
 */
var Login = module.exports = function(config, adapter) {

  if (!(this instanceof Login)) {return new Login(config, adapter); }

  // call super constructor function
  events.EventEmitter.call(this);

  this.config = config;
  this.adapter = adapter;

  // set default routes
  this.loginRoute = config.login.route || '/login';
  var logoutRoute = config.login.logoutRoute || '/logout';

  // change URLs if REST is active
  if (config.rest) {
    this.loginRoute = '/rest' + this.loginRoute;
    logoutRoute = '/rest' + logoutRoute;
  }

  // two-factor authentication route
  this.twoFactorRoute = this.loginRoute + (config.login.twoFactorRoute || '/two-factor');

  var router = new express.Router();
  router.get(this.loginRoute, this.getLogin.bind(this));
  router.post(this.loginRoute, this.postLogin.bind(this));
  router.post(this.twoFactorRoute, this.postTwoFactor.bind(this));
  router.get(logoutRoute, utils.restrict(config), this.getLogout.bind(this));
  this.router = router;

};

util.inherits(Login, events.EventEmitter);



/**
 * GET /login route handling function.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Login.prototype.getLogin = function(req, res, next) {

  var config = this.config;
  var that = this;

  // do not handle the route when REST is active
  if (config.rest) {return next(); }

  // save redirect url
  var suffix = req.query.redirect ? '?redirect=' + encodeURIComponent(req.query.redirect) : '';

  // custom or built-in view
  var view = config.login.views.login || join('get-login');

  // render view
  res.render(view, {
    title: 'Login',
    action: that.loginRoute + suffix,
    basedir: req.app.get('views')
  });
};



/**
 * POST /login route handling function.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Login.prototype.postLogin = function(req, res, next) {

  var adapter = this.adapter;
  var config = this.config;
  var that = this;

  var error = '';

  var login = req.body.login;
  var password = req.body.password;

  // save redirect url
  var suffix = req.query.redirect ? '?redirect=' + encodeURIComponent(req.query.redirect) : '';

  // custom or built-in view
  var view = config.login.views.login || join('get-login');

  // check for valid inputs
  if (!login || !password) {
    error = 'Please enter your email/username and password';

    // send only JSON when REST is active
    if (config.rest) {return res.json(403, {error: error}); }

    // render view
    res.status(403);
    res.render(view, {
      title: 'Login',
      action: that.loginRoute + suffix,
      error: error,
      login: login,
      basedir: req.app.get('views')
    });
    return;
  }

  // check if login is a name or an email address

  // regexp from https://github.com/angular/angular.js/blob/master/src/ng/directive/input.js#L4
  var EMAIL_REGEXP = /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,6}$/;
  var query = EMAIL_REGEXP.test(login) ? 'email' : 'name';

  // find user in db
  adapter.find(query, login, function(err, user) {
    if (err) {return next(err); }

    // no user or user email isn't verified yet -> render error message
    if (!user || !user.emailVerified) {
      error = 'Invalid user or password';

      // send only JSON when REST is active
      if (config.rest) {return res.json(403, {error: error}); }

      // render view
      res.status(403);
      res.render(view, {
        title: 'Login',
        action: that.loginRoute + suffix,
        error: error,
        login: login,
        basedir: req.app.get('views')
      });
      return;
    }

    // check for too many failed login attempts
    if (user.accountLocked && new Date(user.accountLockedUntil) > new Date()) {
      error = 'The account is temporarily locked';

      // send only JSON when REST is active
      if (config.rest) {return res.json(403, {error: error}); }

      // render view
      res.status(403);
      res.render(view, {
        title: 'Login',
        action: that.loginRoute + suffix,
        error: error,
        login: login,
        basedir: req.app.get('views')
      });
      return;
    }

    // if user comes from couchdb it has an 'iterations' key
    if (user.iterations) {pwd.iterations(user.iterations); }

    // compare credentials with data in db
    pwd.hash(password, user.salt, function(hashErr, hash) {
      if (hashErr) {return next(hashErr); }

      if (hash !== user.derived_key) {
        // set the default error message
        var errorMessage = 'Invalid user or password';

        // increase failed login attempts
        user.failedLoginAttempts += 1;

        // lock account on too many login attempts (defaults to 5)
        if (user.failedLoginAttempts >= config.failedLoginAttempts) {
          user.accountLocked = true;

          // set locked time to 20 minutes (default value)
          var timespan = ms(config.accountLockedTime);
          user.accountLockedUntil = moment().add(timespan, 'ms').toDate();

          errorMessage = 'Invalid user or password. Your account is now locked for ' + config.accountLockedTime;
        } else if (user.failedLoginAttempts >= config.failedLoginsWarning) {
          // show a warning after 3 (default setting) failed login attempts
          errorMessage = 'Invalid user or password. Your account will be locked soon.';
        }

        // save user to db
        adapter.update(user, function(updateErr) {
          if (updateErr) {return next(updateErr); }

          // send only JSON when REST is active
          if (config.rest) {return res.json(403, {error: errorMessage}); }

          // send error message
          res.status(403);
          res.render(view, {
            title: 'Login',
            action: that.loginRoute + suffix,
            error: errorMessage,
            login: login,
            basedir: req.app.get('views')
          });
        });

        return;

      }

      // looks like password is correct

      // shift tracking values
      var now = new Date();

      // update previous login time and ip
      user.previousLoginTime = user.currentLoginTime || now;
      user.previousLoginIp = user.currentLoginIp || req.ip;

      // save login time
      user.currentLoginTime = now;
      user.currentLoginIp = req.ip;

      // set failed login attempts to zero but save them in the session
      req.session.failedLoginAttempts = user.failedLoginAttempts;
      user.failedLoginAttempts = 0;
      user.accountLocked = false;

      // save user to db
      adapter.update(user, function(updateErr, updatedUser) {
        if (updateErr) {return next(updateErr); }

        // create session and save the name and email address
        req.session.name = updatedUser.name;
        req.session.email = updatedUser.email;

        // check if two-factor authentication is enabled
        if (!updatedUser.twoFactorEnabled) {

          // get redirect url
          var target = req.query.redirect || '/';

          // user is now logged in
          req.session.loggedIn = true;

          // emit 'login' event
          that.emit('login', updatedUser, res, target);

          // let lockit handle the response
          if (config.login.handleResponse) {
            // send only JSON when REST is active
            if (config.rest) {return res.send(204); }

            // redirect to target url
            res.redirect(target);
          }
          return;
        }

        // two-factor authentication is enabled

        // send only JSON when REST is active
        if (config.rest) {
          return res.json({
            twoFactorEnabled: true
          });
        }

        // custom or built-in view
        var twoFactorView = config.login.views.twoFactor || join('two-factor');

        // render two-factor authentication template
        res.render(twoFactorView, {
          title: 'Two-factor authentication',
          action: that.twoFactorRoute,
          basedir: req.app.get('views')
        });

      });

    });

  });

};



/**
 * POST /login/two-factor.
 *
 * Verify provided token using time-based one-time password.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Login.prototype.postTwoFactor = function(req, res, next) {

  var config = this.config;
  var adapter = this.adapter;
  var loginRoute = this.loginRoute;
  var that = this;

  var token = req.body.token || '';
  var email = req.session.email || '';

  // get redirect url
  var target = req.query.redirect || '/';

  // get user from db
  adapter.find('email', email, function(err, user) {
    if (err) {return next(err); }

    var key = user && user.twoFactorKey;

    // verify POSTed token
    var valid = utils.verify(token, key);

    // redirect to /login if invalid
    if (!valid) {
      // destroy current session
      return utils.destroy(req, function() {
        // send only JSON when REST is active
        if (config.rest) {return res.send(401); }
        res.redirect(loginRoute + '?redirect=' + target);
      });
    }

    // token seems to be fine

    // user is now logged in
    req.session.loggedIn = true;

    // emit 'login' event
    that.emit('login', user, res, target);

    // let lockit handle the response
    if (config.login.handleResponse) {
      // send only JSON when REST is active
      if (config.rest) {return res.send(204); }

      // redirect to target url
      res.redirect(target);
    }

  });

};



/**
 * GET /logout route handling function.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Login.prototype.getLogout = function(req, res) {

  var config = this.config;
  var that = this;

  // save values for event emitter
  var user = {
    name: req.session.name,
    email: req.session.email
  };

  // destroy the session
  utils.destroy(req, function() {
    // clear local variables - they were set before the session was destroyed
    res.locals.name = null;
    res.locals.email = null;

    // emit 'logout' event
    that.emit('logout', user, res);

    // let lockit handle the response
    if (config.login.handleResponse) {

      // send JSON when REST is active
      if (config.rest) {return res.send(204); }

      // custom or built-in view
      var view = config.login.views.loggedOut || join('get-logout');

      // reder logout success template
      res.render(view, {
        title: 'Logout successful',
        basedir: req.app.get('views')
      });

    }
  });

};
